crates/core/component/shielded-pool/src/backref.rs

```diff
@@ -0,0 +1,221 @@
+use anyhow::Result;
+use chacha20poly1305::{
+    aead::{Aead, NewAead},
+    ChaCha20Poly1305, Nonce,
+};
+
+use penumbra_keys::BackreferenceKey;
+use penumbra_sct::Nullifier;
+use penumbra_tct as tct;
+
+pub const ENCRYPTED_BACKREF_LEN: usize = 48;
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Backref {
+    note_commitment: tct::StateCommitment,
+}
+
+#[derive(Clone, Debug)]
+pub struct EncryptedBackref {
+    /// The inner bytes can either have 0 or `ENCRYPTED_BACKREF_LEN` bytes.
+    bytes: Vec<u8>,
+}
+
+impl Backref {
+    pub fn new(note_commitment: tct::StateCommitment) -> Self {
+        Self { note_commitment }
+    }
+
+    pub fn encrypt(
+        &self,
+        brk: &BackreferenceKey,
+        nullifier: &Nullifier,
+    ) -> Result<EncryptedBackref> {
+        let cipher = ChaCha20Poly1305::new(&brk.0);
+
+        // Nonce is the first 12 bytes of the nullifier
+        let nonce_bytes = &nullifier.to_bytes()[..12];
+        let nonce = Nonce::from_slice(&nonce_bytes);
+
+        let plaintext = self.note_commitment.0.to_bytes();
+
+        let ciphertext = cipher
+            .encrypt(nonce, plaintext.as_ref())
+            .map_err(|_| anyhow::anyhow!("encryption error"))?;
+
+        Ok(EncryptedBackref { bytes: ciphertext })
+    }
+}
+
+impl EncryptedBackref {
+    pub fn is_empty(&self) -> bool {
+        self.bytes.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.bytes.len()
+    }
+
+    pub fn dummy() -> Self {
+        Self { bytes: vec![] }
+    }
+
+    /// Decrypts the encrypted backref, returning a backref if the decryption is successful,
+    /// or `None` if the encrypted backref is zero-length.
+    pub fn decrypt(
+        &self,
+        brk: &BackreferenceKey,
+        nullifier: &Nullifier,
+    ) -> Result<Option<Backref>> {
```

Comment:
I think having Result<Option> is smart here, because then you can distinguish between having not put in a back reference, in which case maybe the reference would have pointed to a spend of yours, and a reference you know doesn't belong to you because you can't decrypt it.